### PR TITLE
Base64 decode headers from Confluent Cloud Consume API

### LINKED ISSUE #183: Decode base64 encoded header values returned from CCloud Consume API into UTF-8

### PR DESCRIPTION
**Pull request by rohitsanj**
## Summary of Changes

Resolves #183 

The Confluent Cloud Consume REST API returns base64 encoded header values. Instead of simply returning these as-is to vscode Message Viewer, we process the header values by base64 decoding into UTF-8 strings.

## Any additional details or context that should be provided?

I've simply updated a few test fixtures to expect base64 decoded strings instead.

Before:
Screenshot 2024-11-25 at 2 39 03 PM

After:
Screenshot 2024-11-25 at 2 37 01 PM

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [ ] Added new
    - [x] Updated existing
    - [ ] Deleted existing
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [x] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

### CODE REVIEW

File: src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java [hunk 1]
```diff
@@ -222,6 +225,34 @@ private PartitionConsumeData processPartition(
...
+          String decodedValue;
+          try {
+            decodedValue = new String(
+                BASE64_DECODER.decode(header.value()), StandardCharsets.UTF_8
+            );
+          } catch (IllegalArgumentException e) {
+            // For whatever reason, we couldn't decode the Base64 encoded header value.
+            // Perhaps the API changed or the data is corrupted. Either way,
+            // we log a warning and return the raw value.
+            Log.warn("Failed to base64 decode header value from Confluent Cloud", e);
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:243]: I'm worried about the sidecar spewing this log message w/ stack trace for a lot of messages that all fit the same pattern. In other words, **if** this happens for one record, it's probably going to spew for every record being consumed.

Is it straightforward to count these _per request_, and maybe log it only for the first one (or few)?

Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:243]: Hm true, I think the "only log for first one" will be easy to implement once we have #153 in place.

Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:243]: I'm starting to think that this is probably not a failure case after all. The records may have header values that are not UTF-8 strings.

Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:243]: > I'm starting to think that this is probably not a failure case after all. The records may have header values that are not UTF-8 strings.

What about changing the log level to debug?

File: src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java [hunk 2]
```diff
@@ -222,6 +225,34 @@ private PartitionConsumeData processPartition(
     );
   }
 
+  private static List<SimpleConsumeMultiPartitionResponse.PartitionConsumeRecordHeader>
```
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:228]: nit: I don't think this method needs to be `static`. OTOH, it probably doesn't hurt.

Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:228]: Ah makes sense. I frequently use the `Extract to method` feature in IntelliJ and didn't pay attention to whether it was static or not.

File: src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java (hunk 1 above)
Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java]: Makes sense. Changed log level to debug and added more info to message.

File: src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java [hunk 3]
```diff
@@ -222,6 +225,37 @@ private PartitionConsumeData processPartition(
...
+                BASE64_DECODER.decode(header.value()), StandardCharsets.UTF_8
+            );
+          } catch (IllegalArgumentException e) {
+            // For whatever reason, we couldn't decode the Base64 encoded header value.
+            // Perhaps the API changed or the data is corrupted. Either way,
+            // we log a warning and return the raw value.
+            Log.debugf(e, "Failed to base64 decode header value '%s' from Confluent Cloud" +
+                    "(partition: %d, offset: %d)",
+                header.value(), record.partitionId(), record.offset()
+            );
```
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:246]: uber-nit:

```suggestion
            // we log a debug message and return the raw value.
            Log.debugf(
                e,
                "Failed to base64 decode header value '%s' from Confluent Cloud "
                    + "(partition: %d, offset: %d)",
                header.value(),
                record.partitionId(),
                record.offset()
            );
```

**Review by flippingbits — APPROVED** [review]
Thanks for the update, @rohitsanj! LGTM.

File: src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java (hunk 3 above)
Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/messageviewer/strategy/ConfluentCloudConsumeStrategy.java:246]: Ha, all good. I've addressed it. Thanks!

**Pull request merged**